### PR TITLE
[WIP] Training API

### PR DESCRIPTION
**Pull request by mccorby**
## Affected Dependencies
List any dependencies that are required for this change.

## How has this been tested?
- Describe the tests that you ran to verify your changes.
- Provide instructions so we can reproduce.
- List any relevant details for your test configuration.

## Checklist
- [ ] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [ ] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [ ] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [ ] My changes are covered by tests

### CODE REVIEW

File: demo-app/src/main/java/org/openmined/syft/demo/federated/domain/TrainingTask.kt [hunk 1]
```diff
@@ -39,7 +36,7 @@ class TrainingTask(
                 clientConfig: ClientConfig
             ) {
                 logger.postLog("Model ${model.modelName} received.\n\nStarting training process")
-                trainingProcess(mnistJob, model, plans, clientConfig, logger)
+                mnistJob.train(plans, mnistDataRepository, clientConfig, logger)
```
Comment (vkkhare) [inline demo-app/src/main/java/org/openmined/syft/demo/federated/domain/TrainingTask.kt:39]: We would not need to call train inside the training task. The user will either implement the onReady handler or call scikit learn style train method. Train method cannot subsume all types of training needs and hence it's an optional call foe the user where library implements all the basic iterators etc

File: demo-app/src/main/java/org/openmined/syft/demo/federated/domain/TrainingTask.kt [hunk 2]
```diff
@@ -57,63 +54,11 @@ class TrainingTask(
...
-                val batchSize = (clientConfig.planArgs["batch_size"]
-                                 ?: error("batch_size doesn't exist")).toInt()
-                val batchIValue = IValue.from(
-                    Tensor.fromBlob(longArrayOf(batchSize.toLong()), longArrayOf(1))
-                )
-                val lr = IValue.from(
-                    Tensor.fromBlob(
-                        floatArrayOf(
-                            (clientConfig.planArgs["lr"] ?: error("lr doesn't exist")).toFloat()
-                        ),
```
Comment (vkkhare) [inline demo-app/src/main/java/org/openmined/syft/demo/federated/domain/TrainingTask.kt:88]: Hence we should keep this part of code as sample

File: syftlib/src/main/java/org/openmined/syft/domain/SyftDataRepository.kt [hunk 3]
```diff
@@ -0,0 +1,7 @@
+package org.openmined.syft.domain
+
+import org.pytorch.IValue
+
+interface SyftDataRepository {
```
Comment (vkkhare) [inline syftlib/src/main/java/org/openmined/syft/domain/SyftDataRepository.kt:5]: This part of code will most likely align with @pengyuan dataloader api. We have a basic structure there so may be we can already align both the PR sp they could be merged?

File: syftlib/src/main/java/org/openmined/syft/domain/SyftLogger.kt [hunk 4]
```diff
@@ -1,4 +1,22 @@
...
+    }
+
+    fun postEpoch(epoch: Int) {
+        print(epoch)
+    }
+
+    fun postLog(message: String) {
+        print(message)
+    }
+
```
Comment (vkkhare) [inline syftlib/src/main/java/org/openmined/syft/domain/SyftLogger.kt:15]: Definitely not our thing to do

**Comment by mccorby** [comment]
Superseded by #290

**Pull request closed without merging**